File: sdk/python/gr_results.py
#!/usr/bin/env python3
"""greenpng results SDK (Python) — backend only.

Thin client: get_result / wait_for_result / query. No probe relay; the SDK
only reads analysis results for sessions that were probed by browser/edge.

Usage:
    from gr_results import GrResultClient, cookie_fields

    client = GrResultClient(base_url="https://probe.example.com",
                             api_key=os.environ["GR_SITE_RESULT_KEY"])
    result = client.wait_for_result("sess_...", projection="sdk")
    print(cookie_fields(result))   # {"user_id": "u9", ...} or None

Auth: X-Gr-Sdk-Key with the site-scoped backend key.
Diagnostic projection requires an ops/admin token and is refused by the server for site keys.
"""

from __future__ import annotations

import json
import time
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Dict, Optional


class GrAnalysisPending(Exception):
    """Raised by wait_for_result when the session is still pending at timeout."""

    def __init__(self, last_body: Any):
        self.last_body = last_body
        super().__init__("analysis_pending")


class GrApiError(Exception):
    def __init__(self, message: str, status: Optional[int] = None, body: Any = None):
        self.status = status
        self.body = body
        super().__init__(message)

    @property
    def code(self) -> str:
        if isinstance(self.body, dict):
            err = self.body.get("error") or {}
            if isinstance(err, dict):
                return str(err.get("code") or err.get("message") or "unknown")
            return str(err)
        return "unknown"


class GrResultClient:
    """Read-only result client bound to one probe server + one site key."""

    def __init__(
        self,
        base_url: str,
        api_key: str,
        timeout_ms: int = 15000,
        expected_schema: str = "product_public_v1",
    ):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout_s = timeout_ms / 1000.0
        self.expected_schema = expected_schema

    def _headers(self) -> Dict[str, str]:
        return {
            "Accept": "application/json",
            "X-Gr-Sdk-Key": self.api_key,
            "X-Request-Id": f"req_{int(time.time() * 1000)}",
        }

    def _fetch(self, url: str) -> Any:
        req = urllib.request.Request(url, headers=self._headers())
        try:
            with urllib.request.urlopen(req, timeout=self.timeout_s) as resp:
                return json.loads(resp.read().decode("utf-8") or "{}")
        except urllib.error.HTTPError as e:
            raw = e.read().decode("utf-8", "replace") or "{}"
            try:
                body = json.loads(raw)
            except ValueError:
                body = {"error": raw}
            status = e.code
            if isinstance(body.get("error"), str) and not body["error"].startswith("http_"):
                raise GrApiError(
                    str(body["error"]), status=status, body=body
                ) from None
            raise GrApiError(f"http_{status}", status=status, body=body) from None
        except urllib.error.URLError as e:
            raise GrApiError(str(e.reason)) from None

    def _result_url(
        self,
        session_id: str,
        projection: str,
        strategy_id: Optional[str] = None,
        response_profile: Optional[str] = None,
        profile_cap: Optional[str] = None,
        lang: Optional[str] = None,
    ) -> str:
        params = {"projection": projection}
        for key, value in (
            ("strategy_id", strategy_id),
            ("response_profile", response_profile),
            ("profile_cap", profile_cap),
            ("lang", lang),
        ):
            if value:
                params[key] = value
        return (
            f"{self.base_url}/v1/session/{urllib.parse.quote(session_id, safe='')}"
            f"/result?{urllib.parse.urlencode(params)}"
        )

    def get_result(
        self,
        session_id: str,
        projection: str = "sdk",
        strategy_id: Optional[str] = None,
        response_profile: Optional[str] = None,
        profile_cap: Optional[str] = None,
        lang: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Fetch one result snapshot (projection: public|sdk|diagnostic)."""
        body = self._fetch(self._result_url(
            session_id, projection, strategy_id, response_profile, profile_cap, lang
        ))
        if isinstance(body, dict) and self.expected_schema and \
                body.get("schema_version") and \
                body["schema_version"] != self.expected_schema:
            pass  # tolerate older nodes without schema_version
        return body

    def wait_for_result(
        self,
        session_id: str,
        timeout_ms: int = 8000,
        interval_ms: int = 500,
        projection: str = "sdk",
        strategy_id: Optional[str] = None,
        response_profile: Optional[str] = None,
        profile_cap: Optional[str] = None,
        lang: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Poll until the analysis is no longer pending or timeout_ms elapses."""
        deadline = time.monotonic() + timeout_ms / 1000.0
        last: Any = None
        while True:
            last = self.get_result(
                session_id,
                projection=projection,
                strategy_id=strategy_id,
                response_profile=response_profile,
                profile_cap=profile_cap,
                lang=lang,
            )
            if isinstance(last, dict) and last.get("ok"):
                has_body = bool(last.get("product_public") or last.get("sdk_projection"))
                pending = False
                if isinstance(last.get("product_public"), dict):
                    meta = last["product_public"].get("meta") or {}
                    pending = bool(meta.get("analysis_pending"))
                if has_body and not pending:
                    return last
            if time.monotonic() >= deadline:
                raise GrAnalysisPending(last)
            time.sleep(interval_ms / 1000.0)

    def query(
        self,
        session_id: str,
        projection: str = "sdk",
        wait: bool = False,
        timeout_ms: int = 8000,
        interval_ms: int = 500,
        strategy_id: Optional[str] = None,
        response_profile: Optional[str] = None,
        profile_cap: Optional[str] = None,
        lang: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Query a session result; `wait=True` polls until ready."""
        if wait:
            return self.wait_for_result(
                session_id,
                timeout_ms=timeout_ms,
                interval_ms=interval_ms,
                projection=projection,
                strategy_id=strategy_id,
                response_profile=response_profile,
                profile_cap=profile_cap,
                lang=lang,
            )
        return self.get_result(
            session_id,
            projection=projection,
            strategy_id=strategy_id,
            response_profile=response_profile,
            profile_cap=profile_cap,
            lang=lang,
        )


def cookie_fields(result: Any) -> Optional[Dict[str, str]]:
    """Extract site-allowlisted cookies captured server-side (business ids).

    Reads sdk_projection.cookie_fields, else product_public.cookie_fields,
    else top-level cookie_fields. Returns None when nothing was captured.
    """
    if not isinstance(result, dict):
        return None
    for path in ("sdk_projection", "product_public"):
        node = result.get(path)
        if isinstance(node, dict) and isinstance(node.get("cookie_fields"), dict):
            cf = node["cookie_fields"]
            return cf if cf else None
    if isinstance(result.get("cookie_fields"), dict):
        cf = result["cookie_fields"]
        return cf if cf else None
    return None

File: sdk/go/gr_results.go
// Package grresults is a backend-only, read-only client for greenpng
// analysis results. It does NOT collect or relay browser probes — probes
// keep flowing straight to the probe server; this client only queries
// the stored result per session.
//
// Surface: GetResult / WaitForResult / Query + CookieFields helper.
// Auth: X-Gr-Sdk-Key with the site-scoped backend key.
// Projection: "public" | "sdk" | "diagnostic" (diagnostic needs an ops key).
package grresults

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// AnalysisPending is returned by WaitForResult when the session analysis is
// still pending when the timeout elapses.
type AnalysisPending struct {
	LastBody json.RawMessage
}

func (e *AnalysisPending) Error() string { return "analysis_pending" }

// APIError carries the HTTP status and parsed error body.
type APIError struct {
	Status int
	Body   map[string]interface{}
}

func (e *APIError) Error() string {
	if e.Body != nil {
		if err, ok := e.Body["error"].(string); ok && !strings.HasPrefix(err, "http_") {
			return err
		}
		if errObj, ok := e.Body["error"].(map[string]interface{}); ok {
			if code, ok := errObj["code"].(string); ok {
				return code
			}
		}
	}
	if e.Status > 0 {
		return fmt.Sprintf("http_%d", e.Status)
	}
	return "api_error"
}

// Client is a thin result client for one probe server + one site key.
type Client struct {
	BaseURL string
	APIKey  string
	HTTP    *http.Client

	expectedSchema string
}

// ResultOptions are optional v2 result projection controls. Empty fields use
// the site's configured defaults.
type ResultOptions struct {
	StrategyID      string
	ResponseProfile string
	ProfileCap      string
	Lang            string
}

// New creates a result client. baseURL e.g. "https://probe.example.com".
func New(baseURL, apiKey string) *Client {
	return &Client{
		BaseURL:        strings.TrimRight(baseURL, "/"),
		APIKey:         apiKey,
		HTTP:           &http.Client{Timeout: 15 * time.Second},
		expectedSchema: "product_public_v1",
	}
}

func (c *Client) resultURL(sessionID, projection string, opts ResultOptions) string {
	q := url.Values{}
	q.Set("projection", projection)
	if opts.StrategyID != "" {
		q.Set("strategy_id", opts.StrategyID)
	}
	if opts.ResponseProfile != "" {
		q.Set("response_profile", opts.ResponseProfile)
	}
	if opts.ProfileCap != "" {
		q.Set("profile_cap", opts.ProfileCap)
	}
	if opts.Lang != "" {
		q.Set("lang", opts.Lang)
	}
	return fmt.Sprintf("%s/v1/session/%s/result?%s",
		c.BaseURL, url.PathEscape(sessionID), q.Encode())
}

func (c *Client) fetch(rawURL string) (map[string]interface{}, error) {
	req, err := http.NewRequest(http.MethodGet, rawURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("X-Gr-Sdk-Key", c.APIKey)
	req.Header.Set("X-Request-Id", fmt.Sprintf("req_%d", time.Now().UnixMilli()))
	resp, err := c.HTTP.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	body := map[string]interface{}{}
	if err := json.Unmarshal(raw, &body); err != nil {
		body = map[string]interface{}{"error": fmt.Sprintf("bad_json_%d", resp.StatusCode)}
	}
	if resp.StatusCode >= 400 {
		return body, &APIError{Status: resp.StatusCode, Body: body}
	}
	return body, nil
}

// GetResult fetches one result snapshot for a session.
func (c *Client) GetResult(sessionID, projection string, options ...ResultOptions) (map[string]interface{}, error) {
	var opts ResultOptions
	if len(options) > 0 {
		opts = options[0]
	}
	return c.fetch(c.resultURL(sessionID, projection, opts))
}

// GetResultWithOptions is the explicit v2 form of GetResult.
func (c *Client) GetResultWithOptions(sessionID, projection string, opts ResultOptions) (map[string]interface{}, error) {
	return c.GetResult(sessionID, projection, opts)
}

// WaitForResult polls until the analysis is no longer pending or the timeout
// elapses (AnalysisPending).
func (c *Client) WaitForResult(sessionID, projection string, timeoutMs, intervalMs int, options ...ResultOptions) (map[string]interface{}, error) {
	if timeoutMs <= 0 {
		timeoutMs = 8000
	}
	if intervalMs <= 0 {
		intervalMs = 500
	}
	deadline := time.Now().Add(time.Duration(timeoutMs) * time.Millisecond)
	var last map[string]interface{}
	for {
		last, _ = c.GetResult(sessionID, projection, options...)
		if ok, _ := last["ok"].(bool); ok {
			hasBody := false
			if _, yes := last["product_public"]; yes {
				hasBody = true
			}
			if _, yes := last["sdk_projection"]; yes {
				hasBody = true
			}
			pending := false
			if pp, yes := last["product_public"].(map[string]interface{}); yes {
				if meta, ok := pp["meta"].(map[string]interface{}); ok {
					pending, _ = meta["analysis_pending"].(bool)
				}
			}
			if hasBody && !pending {
				return last, nil
			}
		}
		if time.Now().After(deadline) {
			raw, _ := json.Marshal(last)
			return nil, &AnalysisPending{LastBody: raw}
		}
		time.Sleep(time.Duration(intervalMs) * time.Millisecond)
	}
}

// Query is an alias-style helper: with wait=false it is a single GetResult;
// with wait=true it polls like WaitForResult.
func (c *Client) Query(sessionID, projection string, wait bool, timeoutMs, intervalMs int, options ...ResultOptions) (map[string]interface{}, error) {
	if wait {
		return c.WaitForResult(sessionID, projection, timeoutMs, intervalMs, options...)
	}
	return c.GetResult(sessionID, projection, options...)
}

// CookieFields extracts the site-owner allowlisted cookies captured
// server-side at session open/ingest (business identifiers). Returns nil
// when the site has no cookie allowlist or nothing was captured.
func CookieFields(result map[string]interface{}) (map[string]interface{}, error) {
	look := func(node map[string]interface{}) (map[string]interface{}, bool) {
		cf, ok := node["cookie_fields"].(map[string]interface{})
		if !ok || len(cf) == 0 {
			return nil, false
		}
		return cf, true
	}
	if sp, ok := result["sdk_projection"].(map[string]interface{}); ok {
		if cf, ok := look(sp); ok {
			return cf, nil
		}
	}
	if pp, ok := result["product_public"].(map[string]interface{}); ok {
		if cf, ok := look(pp); ok {
			return cf, nil
		}
	}
	if cf, ok := look(result); ok {
		return cf, nil
	}
	return nil, errors.New("no cookie_fields in result")
}
